internal/pkg/vm/state/state.go

```diff
@@ -0,0 +1,204 @@
+}
+
+func newState(store cbor.IpldStore, root Root, rootNode *hamt.Node) *State {
+	return &State{
+		store:    store,
+		dirty:    false,
+		root:     root,
+		rootNode: rootNode,
+	}
+}
+
+// GetActor retrieves an actor by their key.
+// If the actor is not found it will return false and no error.
+// If there are any IO or decoding errors, it will return false and the error.
+func (st *State) GetActor(ctx context.Context, key actorKey) (*actor.Actor, bool, error) {
+	actorBytes, err := st.rootNode.FindRaw(ctx, string(key.Bytes()))
+	if err == hamt.ErrNotFound {
+		return nil, false, nil
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	var act actor.Actor
+	err = encoding.Decode(actorBytes, &act)
+	if err != nil {
+		return nil, false, err
+	}
+
+	return &act, true, nil
+}
+
+// SetActor sets the the actor to the given value wheter it previously existed or not.
+//
+// This method will not check if the actor previuously existed, it will blindly overwrite it.
+func (st *State) SetActor(ctx context.Context, key actorKey, a *actor.Actor) error {
+	actBytes, err := encoding.Encode(a)
+	if err != nil {
+		return err
+	}
+	if err := st.rootNode.SetRaw(ctx, string(key.Bytes()), actBytes); err != nil {
+		return errors.Wrap(err, "setting actor in state tree failed")
+	}
+	st.dirty = true
+	return nil
+}
+
+// DeleteActor remove the actor from the storage.
+//
+// This method will NOT return an error if the actor was not found.
+func (st *State) DeleteActor(ctx context.Context, key actorKey) error {
+	err := st.rootNode.Delete(ctx, string(key.Bytes()))
+	st.dirty = true
+	if err == hamt.ErrNotFound {
+		return nil
+	}
+	return err
+}
+
+// Commit will flush the state tree into the backing store.
+// The new root is returned.
+func (st *State) Commit(ctx context.Context) (Root, error) {
+	if err := st.rootNode.Flush(ctx); err != nil {
+		return cid.Undef, err
+	}
+
+	// Review: is this put needed because it has not been flushed or because flush does not return the root cid?
+	root, err := st.store.Put(ctx, st.rootNode)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	st.root = root
+	st.dirty = false
+	return st.root, nil
+}
+
+// Root returns the root of the tree and whether that root is dirty.
+func (st *State) Root() (Root, bool) {
+	return st.root, st.dirty
+}
+
+// GetAllActorsResult is the struct returned via a channel by the GetAllActors
+// method. This struct contains only an address string and the actor itself.
+type GetAllActorsResult struct {
+	Key   actorKey
+	Actor *actor.Actor
+	Error error
+}
+
+// GetAllActors returns a channel which provides all actors in the StateTree.
+func (st *State) GetAllActors(ctx context.Context) <-chan GetAllActorsResult {
+	out := make(chan GetAllActorsResult)
+	go func() {
+		defer close(out)
+		st.getActorsFromPointers(ctx, out, st.rootNode.Pointers)
+	}()
+	return out
+}
+
+// NOTE: This extracts actors from pointers recursively. Maybe we shouldn't recurse here.
```

Comment:
I think the depth is very restricted (log N) so it's ok.

Even if you challenge me on using a channel, please use the existing HAMT ForEach method rather than re-implement this deep inspection.

Reply:
inherited code, it's also only used in 1 test and to list the actors in CLI.